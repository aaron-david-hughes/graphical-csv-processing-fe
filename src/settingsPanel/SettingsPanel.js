import React from 'react';
import Popup from "../generalPurposeComponents/popup/Popup";
import FormContentGenerator from "../formContentGenerator/FormContentGenerator";
import Input from "../generalPurposeComponents/input/Input";
import Switch from "../generalPurposeComponents/switch/Switch";
import {processingInputValidityStartState} from "../graphComponentForms/utils/FileNodes";
import Validation from "../graphComponentForms/validation/Validation";
import deepClone from 'lodash.clonedeep';
import Collapsible from "../generalPurposeComponents/collapsible/Collapsible";

class SettingsPanel extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            defaultsEnabled: props.defaultsEnabled,
            config: props.config,
            beforeDefaultConfig: props.config,
            initialConfig: props.initialConfig,
            configChanged: false,
            configValidity: this.setupValidity()
        }
    }

    setupValidity() {
        let validity = {};

        this.props.config.processing.operations.forEach(op => {
            validity[op.operation] = processingInputValidityStartState(op, 'valid');
        });

        return validity;
    }

    async setDefaultsEnabled(e) {
        await this.setState({
            defaultsEnabled: e
        });

        this.props.setDefaultsEnabled(e);
    }

    async textOnBlur(e, operation, field) {
        await this.onBlur(Validation.validateTextField, e, operation, field);
    }

    async numericOnBlur(e, operation, field) {
        await this.onBlur(Validation.validateNumericField, e, operation, field);
    }

    async integerOnBlur(e, operation, field) {
        await this.onBlur(Validation.validateIntegerField, e, operation, field);
    }

    async onBlur(validate, e, operation, field) {
        if (e && e.target && e.target.value !== null && e.target.value.length === 0) {
            let configValidity = deepClone(this.state.configValidity);
            configValidity[operation][field] = 'valid';

            let config = deepClone(this.state.config);
            let opConfig = config.processing.operations.find(op => op.operation === operation);

            opConfig.template[field] = null;

            await this.setState({
                configValidity,
                config
            });
        } else {
            if (validate(e)) {
                let configValidity = deepClone(this.state.configValidity);
                configValidity[operation][field] = 'valid';

                let config = deepClone(this.state.config);
                let opConfig = config.processing.operations.find(op => op.operation === operation);

                opConfig.template[field] = e.target.value;

                await this.setState({
                    configValidity,
                    config
                });
            } else {
                //only mark invalid and change no config state
                let configValidity = deepClone(this.state.configValidity);
                configValidity[operation][field] = 'invalid';

                await this.setState({
                    configValidity
                });
            }
        }

        await this.setState({
            configChanged: this.state.config !== this.state.beforeDefaultConfig
        });
    }

    async dropdownOnChange(e, operation, field) {
        if (Validation.validateTextField(e)) {
            let config = deepClone(this.state.config);
            let opConfig = config.processing.operations.find(op => op.operation === operation);

            opConfig.template[field] = e.target.value;

            await this.setState({
                config
            });
        }
    }

    async switchOnChange(e, operation, field) {
        if (e != null) {
            let config = deepClone(this.state.config);
            let opConfig = config.processing.operations.find(op => op.operation === operation);

            opConfig.template[field] = e;

            await this.setState({
                config
            });
        }
    }

    renderNodeForm(operation, name, template) {
        return <Collapsible
            title={name}
            key={operation}
            open={false}
            id={operation + '-Collapsible'}
            style={{
                marginBottom: '0.5rem',
                marginTop: '0.5rem'
            }}
        >
            <FormContentGenerator
                isSettings={true}
                config={this.state.config}
                inputValidity={this.state.configValidity[operation]}
                nodeOperation={operation}
                nodeObj={template}
                inputActionFunctions={{
                    text: {
                        onBlur: this.textOnBlur.bind(this)
                    },
                    numeric: {
                        onBlur: this.numericOnBlur.bind(this)
                    },
                    integer: {
                        onBlur: this.integerOnBlur.bind(this)
                    },
                    dropdown: {
                        onChange: this.dropdownOnChange.bind(this)
                    },
                    switch: {
                        onChange: this.switchOnChange.bind(this)
                    }
                }}
            />
        </Collapsible>
    }

    render() {
        return <Popup
            id='settingsPanel'
            title='Settings'
            isOpen={true}
            close={this.props.close}
            scrollable={false}
            width='40%'
            height='60%'
        >
            <div
                style={{
                    width: '100%',
                    height: '100%'
                }}
            >
                <div
                    id='settingPanelMainBody'
                    style={{
                        height: `calc(100% - 45px)`,
                        maxHeight: `calc(100% - 45px)`,
                        overflowY: 'scroll'
                    }}
                >
                    <div>
                        <Input
                            id='defaultSwitch'
                            title='Defaults'
                            style={{width: '100%'}}
                            required={false}
                            input={
                                <Switch
                                    id='defaultSwitch-input'
                                    isChecked={this.state.defaultsEnabled}
                                    onChange={async e => this.setDefaultsEnabled(e)}
                                />
                            }
                        />
                    </div>
                    {
                        this.state.defaultsEnabled
                            ? this.state.config.processing.operations.map(op => this.renderNodeForm(op.operation, op.name, op.template))
                            : null
                    }
                </div>
                <div
                    id='settingPanelControls'
                    style={{
                        display: 'inline-flex',
                        marginTop: '10px',
                        paddingTop: '10px',
                        borderTop: '2px solid darkslategray',
                        height: '45px',
                        width: '100%',
                        justifyContent: 'space-between'
                    }}
                >
                    <div id='saveAndCancelButtonDiv'>
                        <button
                            title='save'
                            id='saveButton'
                            disabled={!this.state.defaultsEnabled || !this.state.configChanged}
                            onClick={async () => {
                                await this.props.updateConfig(this.state.config);
                                this.props.close();
                            }}
                        >save</button>
                        <button
                            title='cancel'
                            id='cancelButton'
                            disabled={!this.state.defaultsEnabled}
                            onClick={this.props.close}
                        >cancel</button>
                    </div>
                    <div id='restoreButtonDiv'>
                        <button
                            title='restore'
                            id='restoreButton'
                            disabled={!this.state.defaultsEnabled}
                            onClick={async () => {
                                await this.props.revertConfig();
                                this.props.close();
                            }}
                        >restore default</button>
                    </div>
                </div>
            </div>
        </Popup>
    }
}

export default SettingsPanel;