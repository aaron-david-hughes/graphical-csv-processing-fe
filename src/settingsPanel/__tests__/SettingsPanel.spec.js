import React from 'react';
import {shallow, mount} from 'enzyme';
import SettingsPanel from "../SettingsPanel";
import Config from '../../config.json';

describe('SettingsPanel tests', () => {
    let config = {
        processing: {
            generalTemplate: Config.processing.generalTemplate,
            operations: [
                {
                    operation: 'test',
                    template: {
                        textField: null,
                        numericField: null,
                        integerField: null,
                        dropdownField: 'option1',
                        switchField: true
                    },
                    textField: {
                        input: 'text',
                        title: 'text field title',
                        width: '50%',
                        required: true
                    },
                    numericField: {
                        input: 'numeric',
                        title: 'numeric field title',
                        width: '50%',
                        required: true
                    },
                    integerField: {
                        input: 'integer',
                        title: 'integer field title',
                        width: '50%',
                        required: true
                    },
                    dropdownField: {
                        input: 'dropdown',
                        title: 'dropdown field title',
                        width: '100%',
                        required: true,
                        options: [
                            'option1',
                            'option2'
                        ]
                    },
                    switchField: {
                        input: 'switch',
                        title: 'switch field title',
                        width: '50%',
                        required: true
                    }
                }
            ]
        }
    }

    let props = {
        defaultsEnabled: false,
        initialConfig: config,
        config: config,
        close: jest.fn(),
        updateConfig: jest.fn(),
        revertConfig: jest.fn(),
        setDefaultsEnabled: jest.fn()
    };

    let render;
    let instance;

    beforeEach(() => {
        render = shallow(<SettingsPanel {...props} />);
        instance = render.instance();
    });

    describe('render tests', () => {
        it('should match snapshot when disabled default values', () => {
            expect(render).toMatchSnapshot();
        });

        it('should match snapshot when enabled default values', () => {
            let localProps = {
                ...props,
                defaultsEnabled: true
            };

            render = shallow(<SettingsPanel {...localProps} />);

            expect(render).toMatchSnapshot();
        });

        it('should not run onClick when save button disabled due to defaultsDisabled', async () => {
            render = mount(<SettingsPanel {...props} />);

            await render.find('#saveButton').simulate('click');

            expect(props.updateConfig).not.toHaveBeenCalled();
            expect(props.close).not.toHaveBeenCalled();
        });

        it('should not run onClick when save button disabled due to no config change', async () => {
            let localProps = {
                ...props,
                defaultsEnabled: true
            };

            render = mount(<SettingsPanel {...localProps} />);

            await render.find('#saveButton').simulate('click');

            expect(props.updateConfig).not.toHaveBeenCalled();
            expect(props.close).not.toHaveBeenCalled();
        });

        it('should run onClick when save button enabled', async () => {
            let localProps = {
                ...props,
                defaultsEnabled: true
            };

            render = mount(<SettingsPanel {...localProps} />);
            instance = render.instance();

            instance.setState({
                configChanged: true
            });

            await render.find('#saveButton').simulate('click');

            expect(props.updateConfig).toHaveBeenCalled();
            expect(props.close).toHaveBeenCalled();
        });

        it('should not run onClick when cancel button disabled due to defaultsDisabled', async () => {
            render = mount(<SettingsPanel {...props} />);

            await render.find('#cancelButton').simulate('click');

            expect(props.close).not.toHaveBeenCalled();
        });

        it('should run onClick when cancel button enabled', async () => {
            let localProps = {
                ...props,
                defaultsEnabled: true
            };

            render = mount(<SettingsPanel {...localProps} />);

            await render.find('#cancelButton').simulate('click');

            expect(props.close).toHaveBeenCalled();
        });

        it('should not run onClick when restore defaults button disabled due to defaultsDisabled', async () => {
            render = mount(<SettingsPanel {...props} />);

            await render.find('#restoreButton').simulate('click');

            expect(props.close).not.toHaveBeenCalled();
        });

        it('should run onClick when restore defaults button enabled', async () => {
            let localProps = {
                ...props,
                defaultsEnabled: true
            };

            render = mount(<SettingsPanel {...localProps} />);

            await render.find('#restoreButton').simulate('click');

            expect(props.close).toHaveBeenCalled();
        });
    });

    describe('instance tests', () => {
        describe('setDefaultsEnabled tests', () => {
            it('should set defaultsEnabled and call props setDefaultsEnabled', async () => {
                await instance.setDefaultsEnabled(true);

                expect(instance.state.defaultsEnabled).toEqual(true);
                expect(props.setDefaultsEnabled).toHaveBeenCalled();
            });
        })

        describe('textOnBlur tests', () => {
            it('should set field to be invalid when the default value is invalid', async () => {
                let e = {
                    target: {
                        value: null
                    }
                };

                await instance.textOnBlur(e, 'test', 'textField');

                expect(instance.state.configValidity.test.textField).toEqual('invalid');
                expect(instance.state.config.processing.operations.find(op => op.operation === 'test').template.textField).toEqual(null);
            });

            it('should set field to be valid when the default value is empty', async () => {
                let e = {
                    target: {
                        value: ''
                    }
                };

                await instance.textOnBlur(e, 'test', 'textField');

                expect(instance.state.configValidity.test.textField).toEqual('valid');
                expect(instance.state.config.processing.operations.find(op => op.operation === 'test').template.textField).toEqual(null);
            });

            it('should set field to be valid when the default value is present and valid', async () => {
                let e = {
                    target: {
                        value: 'this is text'
                    }
                };

                await instance.textOnBlur(e, 'test', 'textField');

                expect(instance.state.configValidity.test.textField).toEqual('valid');
                expect(instance.state.config.processing.operations.find(op => op.operation === 'test').template.textField).toEqual('this is text');
            });
        });

        describe('numericOnBlur tests', () => {
            it('should set field to be invalid when the default value is invalid', async () => {
                let e = {
                    target: {
                        value: 'not a number'
                    }
                };

                await instance.numericOnBlur(e, 'test', 'numericField');

                expect(instance.state.configValidity.test.numericField).toEqual('invalid');
                expect(instance.state.config.processing.operations.find(op => op.operation === 'test').template.numericField).toEqual(null);
            });

            it('should set field to be valid when the default value is empty', async () => {
                let e = {
                    target: {
                        value: ''
                    }
                };

                await instance.numericOnBlur(e, 'test', 'numericField');

                expect(instance.state.configValidity.test.numericField).toEqual('valid');
                expect(instance.state.config.processing.operations.find(op => op.operation === 'test').template.numericField).toEqual(null);
            });

            it('should set field to be valid when the default value is present and valid', async () => {
                let e = {
                    target: {
                        value: '1.99'
                    }
                };

                await instance.numericOnBlur(e, 'test', 'numericField');

                expect(instance.state.configValidity.test.numericField).toEqual('valid');
                expect(instance.state.config.processing.operations.find(op => op.operation === 'test').template.numericField).toEqual('1.99');
            });
        });

        describe('integerOnBlur tests', () => {
            it('should set field to be invalid when the default value is invalid', async () => {
                let e = {
                    target: {
                        value: '1.99'
                    }
                };

                await instance.integerOnBlur(e, 'test', 'integerField');

                expect(instance.state.configValidity.test.integerField).toEqual('invalid');
                expect(instance.state.config.processing.operations.find(op => op.operation === 'test').template.integerField).toEqual(null);
            });

            it('should set field to be valid when the default value is empty', async () => {
                let e = {
                    target: {
                        value: ''
                    }
                };

                await instance.integerOnBlur(e, 'test', 'integerField');

                expect(instance.state.configValidity.test.integerField).toEqual('valid');
                expect(instance.state.config.processing.operations.find(op => op.operation === 'test').template.integerField).toEqual(null);
            });

            it('should set field to be valid when the default value is present and valid', async () => {
                let e = {
                    target: {
                        value: '1'
                    }
                };

                await instance.integerOnBlur(e, 'test', 'integerField');

                expect(instance.state.configValidity.test.integerField).toEqual('valid');
                expect(instance.state.config.processing.operations.find(op => op.operation === 'test').template.integerField).toEqual('1');
            });
        });

        describe('dropdownOnChange tests', () => {
            it('should not set template value to input if invalid', async () => {
                let priorValidity = instance.state.configValidity.test.dropdownField;
                let priorValue = instance.state.config.processing.operations.find(op => op.operation === 'test').template.dropdownField;

                let e = {
                    target: {
                        value: null
                    }
                };

                await instance.dropdownOnChange(e, 'test', 'dropdownField');

                expect(instance.state.configValidity.test.dropdownField).toEqual(priorValidity);
                expect(instance.state.config.processing.operations.find(op => op.operation === 'test').template.dropdownField).toEqual(priorValue);
            });

            it('should set template value to input if valid', async () => {
                let e = {
                    target: {
                        value: 'option2'
                    }
                };

                await instance.dropdownOnChange(e, 'test', 'dropdownField');

                expect(instance.state.configValidity.test.dropdownField).toEqual('valid');
                expect(instance.state.config.processing.operations.find(op => op.operation === 'test').template.dropdownField).toEqual('option2');
            });
        });

        describe('switchOnChange tests', () => {
            it('should not set template value to input if invalid', async () => {
                let priorValidity = instance.state.configValidity.test.switchField;
                let priorValue = instance.state.config.processing.operations.find(op => op.operation === 'test').template.switchField;

                let e = null;

                await instance.switchOnChange(e, 'test', 'switchField');

                expect(instance.state.configValidity.test.switchField).toEqual(priorValidity);
                expect(instance.state.config.processing.operations.find(op => op.operation === 'test').template.switchField).toEqual(priorValue);
            });

            it('should set template value to input if valid', async () => {
                let e = false;

                await instance.switchOnChange(e, 'test', 'switchField');

                expect(instance.state.configValidity.test.switchField).toEqual('valid');
                expect(instance.state.config.processing.operations.find(op => op.operation === 'test').template.switchField).toEqual(false);
            });
        });
    });
});