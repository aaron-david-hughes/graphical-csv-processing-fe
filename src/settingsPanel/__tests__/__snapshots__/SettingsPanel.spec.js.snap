// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsPanel tests render tests should match snapshot when disabled default values 1`] = `
<Popup
  close={[MockFunction]}
  height="60%"
  id="settingsPanel"
  isOpen={true}
  scrollable={false}
  title="Settings"
  width="40%"
>
  <div
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  >
    <div
      id="settingPanelMainBody"
      style={
        Object {
          "height": "calc(100% - 45px)",
          "maxHeight": "calc(100% - 45px)",
          "overflowY": "scroll",
        }
      }
    >
      <div>
        <Input
          id="defaultSwitch"
          input={
            <Switch
              id="defaultSwitch-input"
              isChecked={false}
              onChange={[Function]}
            />
          }
          required={false}
          style={
            Object {
              "width": "100%",
            }
          }
          title="Defaults"
        />
      </div>
    </div>
    <div
      id="settingPanelControls"
      style={
        Object {
          "borderTop": "2px solid darkslategray",
          "display": "inline-flex",
          "height": "45px",
          "justifyContent": "space-between",
          "marginTop": "10px",
          "paddingTop": "10px",
          "width": "100%",
        }
      }
    >
      <div
        id="saveAndCancelButtonDiv"
      >
        <button
          disabled={true}
          id="saveButton"
          onClick={[Function]}
          title="save"
        >
          save
        </button>
        <button
          disabled={true}
          id="cancelButton"
          onClick={[MockFunction]}
          title="cancel"
        >
          cancel
        </button>
      </div>
      <div
        id="restoreButtonDiv"
      >
        <button
          disabled={true}
          id="restoreButton"
          onClick={[Function]}
          title="restore"
        >
          restore default
        </button>
      </div>
    </div>
  </div>
</Popup>
`;

exports[`SettingsPanel tests render tests should match snapshot when enabled default values 1`] = `
<Popup
  close={[MockFunction]}
  height="60%"
  id="settingsPanel"
  isOpen={true}
  scrollable={false}
  title="Settings"
  width="40%"
>
  <div
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  >
    <div
      id="settingPanelMainBody"
      style={
        Object {
          "height": "calc(100% - 45px)",
          "maxHeight": "calc(100% - 45px)",
          "overflowY": "scroll",
        }
      }
    >
      <div>
        <Input
          id="defaultSwitch"
          input={
            <Switch
              id="defaultSwitch-input"
              isChecked={true}
              onChange={[Function]}
            />
          }
          required={false}
          style={
            Object {
              "width": "100%",
            }
          }
          title="Defaults"
        />
      </div>
      <Collapsible
        id="test-Collapsible"
        key="test"
        open={false}
        style={
          Object {
            "marginBottom": "0.5rem",
            "marginTop": "0.5rem",
          }
        }
      >
        <FormContentGenerator
          config={
            Object {
              "processing": Object {
                "generalTemplate": Object {
                  "group": "processing",
                  "hovered": Object {
                    "fill": "blue",
                    "shape": "diamond",
                  },
                  "normal": Object {
                    "fill": "blue",
                    "shape": "diamond",
                  },
                  "selected": Object {
                    "fill": "blue",
                    "shape": "diamond",
                  },
                },
                "operations": Array [
                  Object {
                    "dropdownField": Object {
                      "input": "dropdown",
                      "options": Array [
                        "option1",
                        "option2",
                      ],
                      "required": true,
                      "title": "dropdown field title",
                      "width": "100%",
                    },
                    "integerField": Object {
                      "input": "integer",
                      "required": true,
                      "title": "integer field title",
                      "width": "50%",
                    },
                    "numericField": Object {
                      "input": "numeric",
                      "required": true,
                      "title": "numeric field title",
                      "width": "50%",
                    },
                    "operation": "test",
                    "switchField": Object {
                      "input": "switch",
                      "required": true,
                      "title": "switch field title",
                      "width": "50%",
                    },
                    "template": Object {
                      "dropdownField": "option1",
                      "integerField": null,
                      "numericField": null,
                      "switchField": true,
                      "textField": null,
                    },
                    "textField": Object {
                      "input": "text",
                      "required": true,
                      "title": "text field title",
                      "width": "50%",
                    },
                  },
                ],
              },
            }
          }
          inputActionFunctions={
            Object {
              "dropdown": Object {
                "onChange": [Function],
              },
              "integer": Object {
                "onBlur": [Function],
              },
              "numeric": Object {
                "onBlur": [Function],
              },
              "switch": Object {
                "onChange": [Function],
              },
              "text": Object {
                "onBlur": [Function],
              },
            }
          }
          inputValidity={
            Object {
              "dropdownField": "valid",
              "integerField": "valid",
              "numericField": "valid",
              "switchField": "valid",
              "textField": "valid",
            }
          }
          isSettings={true}
          nodeObj={
            Object {
              "dropdownField": "option1",
              "integerField": null,
              "numericField": null,
              "switchField": true,
              "textField": null,
            }
          }
          nodeOperation="test"
        />
      </Collapsible>
    </div>
    <div
      id="settingPanelControls"
      style={
        Object {
          "borderTop": "2px solid darkslategray",
          "display": "inline-flex",
          "height": "45px",
          "justifyContent": "space-between",
          "marginTop": "10px",
          "paddingTop": "10px",
          "width": "100%",
        }
      }
    >
      <div
        id="saveAndCancelButtonDiv"
      >
        <button
          disabled={true}
          id="saveButton"
          onClick={[Function]}
          title="save"
        >
          save
        </button>
        <button
          disabled={false}
          id="cancelButton"
          onClick={[MockFunction]}
          title="cancel"
        >
          cancel
        </button>
      </div>
      <div
        id="restoreButtonDiv"
      >
        <button
          disabled={false}
          id="restoreButton"
          onClick={[Function]}
          title="restore"
        >
          restore default
        </button>
      </div>
    </div>
  </div>
</Popup>
`;
